### PR TITLE
Adding ROCm support for MLIR generated kernels (in the non-XLA path)

### PR DESCRIPTION
**Pull request by deven-amd**
This PR adds ROCm support for MLIR generated kernels in the non-XLA path.

As part of this implementation, this PR also
* changes the way `amdgpu_targets` are determined for the TF build (see first commit)
* adds `clang-offload-bundler` and `bin2c.py` to the `local_config_rocm` repo (see second and third commits)

The last commit makes "MLIR generated kernels" the default in the ROCM build. That is for an initial CI run only (to see if it results in regressions...the corresponding unit tests for "MLIR generated kernels" have already been confirmed to pass). It (the last commit) will be removed before the PR is merged

/cc @whchung

### CODE REVIEW

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

File: tensorflow/compiler/xla/service/hlo_module_config.cc [hunk 1]
```diff
@@ -51,12 +51,14 @@ string HloModuleConfig::compilation_cache_key() const {
   string key = absl::StrCat("profiling=", hlo_profiling_enabled());
   StrAppend(&key, "::(");
   std::vector<string> params;
-  for (const ShapeLayout& param_layout :
-       entry_computation_layout_->parameter_layouts()) {
-    params.push_back(param_layout.shape().DebugString());
+  if (entry_computation_layout_.has_value()) {
```
Comment (whchung) [inline tensorflow/compiler/xla/service/hlo_module_config.cc:54]: @deven-amd what's the rationale behind this change?

File: third_party/gpus/rocm_configure.bzl [hunk 2]
```diff
@@ -612,6 +609,19 @@ def _create_local_rocm_repository(repository_ctx):
         outs = rocm_lib_outs,
     ))
 
+    # copy files mentioned in third_party/nccl/build_defs.bzl.tpl
```
Comment (whchung) [inline third_party/gpus/rocm_configure.bzl]: nit: `nccl`?

File: third_party/gpus/rocm_configure.bzl [hunk 3]
```diff
@@ -718,10 +729,6 @@ def _create_local_rocm_repository(repository_ctx):
             "%{hip_runtime_path}": rocm_config.rocm_toolkit_path + "/hip/lib",
             "%{hip_runtime_library}": "amdhip64",
             "%{crosstool_verbose}": _crosstool_verbose(repository_ctx),
-            "%{gcc_host_compiler_path}": str(cc),
```
Comment (whchung) [inline third_party/gpus/rocm_configure.bzl:721]: `gcc_host_compiler_path` is an obsolete variable so it's removed here, correct?

File: tensorflow/core/kernels/mlir_generated/build_defs.bzl [hunk 4]
```diff
@@ -15,9 +22,12 @@ def _lookup_file(filegroup, path):
             return file
     return None
 
-CubinInfo = provider(fields = ["cubins"])
+CubinInfo = provider(fields = [
```
Comment (whchung) [inline tensorflow/core/kernels/mlir_generated/build_defs.bzl]: could `CubinInfo` be renamed?

File: third_party/gpus/rocm_configure.bzl (hunk 2 above)
Comment (deven-amd) [reply third_party/gpus/rocm_configure.bzl]: copy-paste error, will fix

File: third_party/gpus/rocm_configure.bzl (hunk 3 above)
Comment (deven-amd) [reply third_party/gpus/rocm_configure.bzl:721]: correct.
`gcc_host_compiler_path` is no longer referenced/used in the `crosstool_wrapper_driver_rocm` and hence removing it here as well

File: tensorflow/core/kernels/mlir_generated/build_defs.bzl (hunk 4 above)
Comment (deven-amd) [reply tensorflow/core/kernels/mlir_generated/build_defs.bzl]: Yes, it should be. will rename it

File: tensorflow/compiler/xla/service/hlo_module_config.cc (hunk 1 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/hlo_module_config.cc:54]: `entry_computation_layout_` does not seem to have any value associated with it, for the llvm module that is generated from the MLIR generated kernel. 

For the ROCm flow, the XLA backend will call the `HloModuleConfig::compilation_cache_key()` when determining the hash value of the llvm module (as part of the HSACOCache implementation). This call will crash, without the `if` guard being added here

**Comment by deven-amd** [comment]
links to the passing CI logs with "MLIR Generated GPU kernels" enabled by default

* http://ml-ci.amd.com:21096/job/develop-upstream-unit-tests/job/tensorflow-upstream-unit-tests/2114/
* http://ml-ci.amd.com:21096/job/develop-upstream-unit-tests/job/tensorflow-upstream-unit-tests-rocm-cpp/676/
* http://ml-ci.amd.com:21096/job/develop-upstream-unit-tests/job/tensorflow-upstream-unit-tests-rocm-xla/1468/

My next rebase and push will disable (by default) the use of "MLIR Generates GPU kernels"
